src/operators/transfermatrix.jl

```diff
@@ -0,0 +1,234 @@
+"""
+function InfiniteTransferPEPO(T::InfinitePEPS, O::InfinitePEPO, dir, row)
+    T = rotate_north(T, dir)
+    O = rotate_north(O, dir)
+    return InfiniteTransferPEPO(PeriodicArray(T[row, :]), PeriodicArray(O[row, :, :]))
+end
+
+Base.size(transfer::InfiniteTransferPEPO) = size(transfer.top)
+Base.size(transfer::InfiniteTransferPEPO, args...) = size(transfer.top, args...)
+Base.length(transfer::InfiniteTransferPEPO) = size(transfer, 1)
+height(transfer::InfiniteTransferPEPO) = size(transfer.mid, 2)
+Base.getindex(O::InfiniteTransferPEPO, i) = (O.top[i], O.bot[i], Tuple(O.mid[i, :]))
+
+Base.iterate(O::InfiniteTransferPEPO, i=1) = i > length(O) ? nothing : (O[i], i + 1)
+
+VectorInterface.scalartype(::InfiniteTransferPEPO{T}) where {T} = scalartype(T)
+
+function virtual_space(O::InfiniteTransferPEPO, i, dir)
+    return prod([
+        space(O.top[i], dir + 1),
+        space.(O.mid[i, :], Ref(dir + 2))...,
+        space(O.bot[i], dir + 1)',
+    ])
+end
+
+"""
+    const MultilineTransferPEPO = MPSKit.Multiline{<:InfiniteTransferPEPO}
+
+Type that represents a multi-line transfer operator, where each line each corresponds to a
+row of a partition function encoding the overlap of an `InfinitePEPO` between 'ket' and
+'bra' `InfinitePEPS` states.
+"""
+const MultilineTransferPEPO = MPSKit.Multiline{<:InfiniteTransferPEPO}
+Base.convert(::Type{MultilineTransferPEPO}, O::InfiniteTransferPEPO) = MPSKit.Multiline([O])
+
+"""
+    MultilineTransferPEPO(T::InfinitePEPS, O::InfinitePEPO, dir)
+
+Construct a multi-row transfer operator corresponding to the partition function representing
+the expectation value of `O` for the state `T`. The partition function is first rotated such
+that the direction `dir` faces north.
+"""
+function MultilineTransferPEPO(T::InfinitePEPS, O::InfinitePEPO, dir)
+    rowsize = size(T, mod1(dir, 2))  # depends on dir
+    return MPSKit.Multiline(map(cr -> InfiniteTransferPEPO(T, O, dir, cr), 1:rowsize))
+end
+
+#
+# Common interface
+#
+
+const InfiniteTransferMatrix = Union{InfiniteTransferPEPS,InfiniteTransferPEPO}
+const MultilineTransferMatrix = Union{MultilineTransferPEPS,MultilineTransferPEPO}
+
+_elementwise_dual(S::ElementarySpace) = S
+_elementwise_dual(P::ProductSpace) = prod(dual.(P))
+
+north_space(O::InfiniteTransferMatrix, i) = virtual_space(O, i, NORTH)
+east_space(O::InfiniteTransferMatrix, i) = virtual_space(O, i, EAST)
+south_space(O::InfiniteTransferMatrix, i) = virtual_space(O, i, SOUTH)
+west_space(O::InfiniteTransferMatrix, i) = virtual_space(O, i, WEST)
+
+MPSKit.left_virtualspace(O::InfiniteTransferMatrix, i) = west_space(O, i)
+function MPSKit.right_virtualspace(O::InfiniteTransferMatrix, i)
+    return _elementwise_dual(east_space(O, i))
+end # follow MPSKit convention: right vspace gets a dual by default
+
+Base.getindex(t::MultilineTransferMatrix, i::Colon, j::Int) = Base.getindex.(t.data[i], j)
+Base.getindex(t::MultilineTransferMatrix, i::Int, j) = Base.getindex(t.data[i], j)
```

Comment:
Are these necessary? In principle this should already be correctly dispatched by `Multiline`

Reply:
I don't think so? `Multiline` only has a `Base.getindex(m::Multiline, i::Int) = getindex(parent(m), i)`. These ones are the equivalent of the ones defined for `MultilineMPO`:
```julia
Base.getindex(t::MultilineMPO, ::Colon, j::Int) = Base.getindex.(t.data, j)
Base.getindex(t::MultilineMPO, i::Int, j) = Base.getindex(t[i], j)
```
So I have to define them since I can't subtype `MultilineMPO`. I'll clean them up a bit though.

Reply:
I rewrote things to use `parent(t)` instead of `t.data`, which is the cleanest thing I can do right now I think. I don't think it makes sense to define these for generic `Multiline`s though, since this really assumes each row can in turn be indexed in a sensible way and this is not generally the case.

Reply:
Sorry my bad, somehow I was already assuming that these were already `MultilineMPO` objects. I would be okay having this as a generic definition in MPSKit as well, but let's keep this as is for now.